src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/EETypeNode.cs

```diff
@@ -600,14 +629,67 @@ public override bool HasDynamicDependencies
                 // This node's EETypeOptionalFields node may change if this EEType implements interfaces
                 // that are used since the dispatch map table index is computed once we know the interface
                 // layout later on in compilation.
-                return _type.RuntimeInterfaces.Length > 0 && _constructed;
+
+                // Also, if any of the implemented interfaces have variance, variant dispatch might result in
+                // slots being necessary.
+                foreach (var intfType in _type.GetClosestMetadataType().RuntimeInterfaces)
```

Comment:
Does this need to be in `HasDynamicDependencies`?  It's more building up state than answering the question. Also we should guard against multiple calls if we leave it here so we don't add to `_variantInterfacesImplemented` multiple times.